File: flut/flutter/animation/animation.py
from typing import Callable

from flut._flut.engine import FlutEnumObject
from flut.flutter.foundation.change_notifier import Listenable


class AnimationStatus(FlutEnumObject):
    dismissed: "AnimationStatus"
    forward: "AnimationStatus"
    reverse: "AnimationStatus"
    completed: "AnimationStatus"

    @property
    def isAnimating(self) -> bool:
        return self is AnimationStatus.forward or self is AnimationStatus.reverse

    @property
    def isCompleted(self) -> bool:
        return self is AnimationStatus.completed

    @property
    def isDismissed(self) -> bool:
        return self is AnimationStatus.dismissed

    @property
    def isForwardOrCompleted(self) -> bool:
        return self is AnimationStatus.forward or self is AnimationStatus.completed


class Animation(Listenable):
    _flut_type = "Animation"

    @property
    def value(self):
        raise NotImplementedError

    @property
    def status(self) -> AnimationStatus:
        raise NotImplementedError

    @property
    def isAnimating(self) -> bool:
        return (
            self.status is AnimationStatus.forward
            or self.status is AnimationStatus.reverse
        )

    @property
    def isCompleted(self) -> bool:
        return self.status is AnimationStatus.completed

    @property
    def isDismissed(self) -> bool:
        return self.status is AnimationStatus.dismissed

    @property
    def isForwardOrCompleted(self) -> bool:
        return (
            self.status is AnimationStatus.forward
            or self.status is AnimationStatus.completed
        )

    def addStatusListener(self, listener: Callable[[AnimationStatus], None]):
        self._flut_call_with_callable(
            "addStatusListener", listener, "AnimationStatusListener"
        )

    def removeStatusListener(self, listener: Callable[[AnimationStatus], None]):
        self._flut_remove_callable(
            "removeStatusListener", "addStatusListener", listener
        )

File: flut/.flutter/lib/flutter/animation/animation.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flutter/foundation/change_notifier.dart';

class FlutAnimationStatus extends FlutEnumObject<AnimationStatus> {
  const FlutAnimationStatus()
    : super('AnimationStatus', const {
        'dismissed': AnimationStatus.dismissed,
        'forward': AnimationStatus.forward,
        'reverse': AnimationStatus.reverse,
        'completed': AnimationStatus.completed,
      });
}

class FlutAnimation<T extends Animation<dynamic>> extends FlutListenable<T> {
  final Map<int, AnimationStatusListener> statusListenersByCid = {};

  FlutAnimation.createFromData({
    required super.runtime,
    required super.data,
    required super.target,
  }) : super.createFromData();

  FlutAnimation.createFromObject({
    required super.runtime,
    required super.oid,
    required super.type,
    required super.target,
  }) : super.createFromObject();

  void clearTrackedStatusListeners() {
    for (final cb in statusListenersByCid.values) {
      flutTarget.removeStatusListener(cb);
    }
    statusListenersByCid.clear();
  }

  @override
  dynamic getRawProperty(String property) {
    switch (property) {
      case 'value':
        return flutTarget.value;
      case 'status':
        return flutTarget.status;
    }
    return super.getRawProperty(property);
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    switch (method) {
      case 'addStatusListener':
        final cb = args[0] as AnimationStatusListener;
        flutTarget.addStatusListener(cb);
        final cid = runtime.cidForCallable(cb);
        if (cid != null) statusListenersByCid[cid] = cb;
        return null;
      case 'removeStatusListener':
        final cidArg = args[0];
        if (cidArg is int) {
          final cb = statusListenersByCid.remove(cidArg);
          if (cb != null) flutTarget.removeStatusListener(cb);
        }
        return null;
    }
    return super.callMethod(method, args, kwargs);
  }
}
